File: js/chris.js
const fs = require('fs');
const { performance } = require('perf_hooks');

const input = fs.readFileSync('../input.txt',
  { encoding: 'utf8', flag: 'r' });

function findFirstFourteen(findString, window) {
  let state = 0;

  let i = window - 1;
  while (i < findString.length) {
    const end = (i + 1) - window;

    let w = i;
    while (w >= end) {
      const marker = 1 << (findString.charCodeAt(w) & 31);

      if ((state & marker) === 0) {
        state = state | marker;
      } else {
        i = w + window;
        state = 0;
        break;
      }

      if (w === end) {
        return i + 1;
      }

      w--;
    }
  }

  return 0;
}

function callFindFirstFourteen() {
  return findFirstFourteen(input, 14);
}

%PrepareFunctionForOptimization(callFindFirstFourteen);

let start = 0;

// Call function once to fill type information
start = callFindFirstFourteen();

// Call function again to go from uninitialized -> pre-monomorphic -> monomorphic
start = callFindFirstFourteen();

%OptimizeFunctionOnNextCall(callFindFirstFourteen);

start = callFindFirstFourteen();

let t = 100;
const times = [];

while (t > 0) {
  const before = performance.now();
  start = callFindFirstFourteen();
  const after = performance.now();

  times.push((after - before) * 1000000);

  t--;
}

const average = times.reduce((a, b) => a + b) / times.length;

console.log(`findFirstFourteen found ${start} and took ${average} nanoseconds`);

File: zig/chris.zig
const std = @import("std");
const data = @embedFile("input.txt");

fn findFirstFourteen(findString: *const [4095:0]u8, window: u8) u64 {
    const one: u32 = 1;

    var check: u32 = 0;
    var count: u32 = 0;

    var i: u64 = window - 1;
    while (i < findString.len) {
        const end = (i + 1) - window;

        var w: u64 = i;
        while (w >= end) {
            const marker: u32 = one << @truncate(findString[w] & 31);

            if ((check & marker) == 0) {
                check = check | marker;
                count += 1;
            } else {
                i = w + window;
                check = 0;
                count = 0;
                break;
            }

            if (count == window) {
                return i + 1;
            }

            w -= 1;
        }
    }

    return 0;
}

pub fn main() !void {
    const before = std.time.nanoTimestamp();
    const start = findFirstFourteen(data, 14);
    const after = std.time.nanoTimestamp();

    const timeTaken = after - before;

    std.debug.print("findStart found {d} and took {d} nanoseconds\n", .{ start, timeTaken });
}
